crates/models/codegen/src/lib.rs

```diff
@@ -0,0 +1,427 @@
+                0,
+                2,
+                1,
+                3,
+            );
+
+            let kq = ctx0.op_mul_mat(&big_k, &q);
+            let kq_scaled = ctx0.op_scale(
+                &kq,
+                &ctx0.new_f32(1f32 / f32::sqrt(n_embd as f32 / n_head as f32)),
+            );
+
+            let kq_masked = ctx0.op_diag_mask_inf(&kq_scaled, n_past);
+            let kq_softmax = ctx0.op_soft_max(&kq_masked);
+
+            let big_v = ctx0.op_view_3d(
+                memory_v,
+                (n_past + n, n_embd / n_head, n_head),
+                (
+                    n_ctx * memory_v_size,
+                    n_ctx * memory_v_size * n_embd / n_head,
+                ),
+                il * n_ctx * memory_v_size * n_embd,
+            );
+
+            let kqv = ctx0.op_mul_mat(&big_v, &kq_softmax);
+            let kqv_merged = ctx0.op_permute(&kqv, 0, 2, 1, 3);
+
+            current = ctx0.op_cpy(&kqv_merged, &ctx0.new_tensor_2d(ggml::Type::F32, n_embd, n));
+
+            // self-attention projection
+            current = ctx0.op_mul_mat(&self.layers[il].c_attn_proj_w, &current);
+
+            // feed-forward
+            let ff_in = current.share();
+
+            current = ctx0.op_mul_mat(&self.layers[il].c_mlp_fc_w, &input_sa);
+            current = ctx0.op_add(
+                &ctx0.op_repeat(&self.layers[il].c_mlp_fc_b, &current),
+                &current,
+            );
+
+            current = ctx0.op_gelu(&current);
+
+            // feed-forward projection
+            current = ctx0.op_mul_mat(&self.layers[il].c_mlp_proj_w, &current);
+            current = ctx0.op_add(
+                &ctx0.op_repeat(&self.layers[il].c_mlp_proj_b, &current),
+                &current,
+            );
+
+            current = ctx0.op_add(&current, &ff_in);
+
+            // input for next layer
+            input_layer = ctx0.op_add(&current, &input_layer);
+        }
+
+        // norm
+        input_layer = ctx0.op_norm(&input_layer);
+        input_layer = ctx0.op_add(
+            &ctx0.op_mul(&ctx0.op_repeat(&self.ln_f_g, &input_layer), &input_layer),
+            &ctx0.op_repeat(&self.ln_f_b, &input_layer),
+        );
+
+        // lm_head
+        input_layer = ctx0.op_mul_mat(&self.lmh_g, &input_layer);
+        input_layer = ctx0.op_add(&ctx0.op_repeat(&self.lmh_b, &input_layer), &input_layer);
+
+        // run the computation
+        gf.build_forward_expand(&input_layer);
+        ctx0.graph_compute(&mut gf);
+
+        // finish evaluation
+        common::read_last_token(session, &input_layer, n_vocab, n);
+        common::extract_logits(output_request, &input_layer, n_vocab, n);
+        common::extract_embeddings(output_request, &embd, n_embd, n);
+        common::update_session(session, &ctx0, input_tokens.len(), n);
+    }
+
+    fn vocabulary(&self) -> &Vocabulary {
+        &self.vocabulary
+    }
+
+    fn n_context_tokens(&self) -> usize {
+        self.hyperparameters.n_ctx
```

Comment:
Is this right? I'm worried this is causing problems with GPT-2, but I'm not sure...
https://github.com/rustformers/llm/issues/167#issuecomment-1537160402

Reply:
probably not I'll fix that when refactoring. Goal is just to get it working with a converted codegen model